Add support for wave extensible format

dtruong27 opened the pull request:
Hello! Thank you very much for this nice framework!

We have some wave files obtained from our iOS app where we store the recording data directly from the microphone to file, without any processing or resampling applied. The iOS stores the files in the wave extensible format. Therefore, we could not load them with this framework in our flutter app.

https://stackoverflow.com/questions/29000788/large-wave-file-not-being-read-in-python

This pull request is to support the extensible format. The wave file added in the test data was obtained from our iOS app.

Thank you very much in advance!

liamappelbe commented:
Supporting this format sounds like a great idea. A few high level comments:

- Do you know if there's a spec somewhere, or maybe a reference reader implementation we can look at? I'm a bit suspicious that the change is so simple. In particular, I'd be surprised if `WAVE_FORMAT_EXTENSIBLE` only supported 32 and 64 bit floats. I think something more must be going on. When I google around a bit, I'm seeing something about this format adding extra metadata fields, so the `read` function is probably going to need to change too.
- It's not ideal to reuse the `WavFormat.float32/64` formats. Better to add new values to that enum for the variants of this new format.
- I'd like to see some more in-depth tests. You're test verifies a few of the metadata fields, but not the samples that were read. So if it turns out that the format is more complicated, and the samples are wrong, your test won't catch it. You're also only testing the 32-bit case, so you'll need a 64-bit test too. I recommend getting some test data with known samples from an unrelated source and testing that the samples match what you expect.
  - Does Audacity or some other audio program support this format? If so you could try generating a sine wave, export it, then load it in your test, and verify that the samples match Dart's sine function. (I did this for the other formats when I was first writing this package, but it seems I deleted all the old test data and just kept these white noise ones. Probably a mistake in hindsight 😅)
  - If you know of a package in some other language, eg python, that supports *writing* these files, you could generate a reference file that way.
- Please also bump the package version to 1.4.0 in the pubspec.yaml, and add a changelog entry, so I can release it when it's ready.
- Not related to your change, but the github CI bot doesn't seem to be running any tests. Not sure why. I'll need to investigate so we can run the tests before this lands. Looks like the build status badge is also broken.

liamappelbe commented:
Ok, I've re-enabled the workflow, and it runs on my [test PR](https://github.com/liamappelbe/wav/pull/18), but I don't see a way of forcing it to run on this PR. It's supposed to run whenever you push, so maybe it'll start running when you push your next commit to this PR?

dtruong27 commented:
Hi, thanks a lot for the review and the suggestion for changes. I tried to address as much as possible what you proposed, but there might be something missing or not as you expected. Please check and let me know.

liamappelbe commented:
I read the reference you linked. It says "`WAVE_FORMAT_EXTENSIBLE` format should be used whenever":
- "PCM data has more than 16 bits/sample"
  - but I've definitely seen ordinary wavs with 24/32 bits/sample, and this package already supports this
- "The number of channels is more than 2"
  - but the channels field is a uint16, so there's no reason an ordinary wav couldn't have more channels. This package can handle this too
- "The actual number of bits/sample is not equal to the container size"
  - Sounds reasonable to support for reading, but I don't really see a need to support this in writing, since it doesn't reduce the file size at all
  - For reading support, it's not clear from the documentation whether it's necessary to worry about this field though, since it also says "The number of valid bits (`wValidBitsPerSample`) is informational only. The data is correctly represented in the precision of the container size."
  - Maybe we'd need to take this field into account when converting the samples to doubles? But I'd want to see some real-world wavs that have this field set to a value less than the container size to be sure. For example, if we have an 8-bit container but only 7 valid bits, do the samples go from 0 to 127, or do they use the full range but only even numbers? Let me know if you have a real file we can take a look at.
- "The mapping from channels to speakers needs to be specified"
  - Doesn't seem relevant to me, unless a user has a need for it. I also don't understand how this field is encoded.
  - Does your use case require this field?

If we don't have a use case for writing support for any of these fields, then I think we probably don't need a new `WavFormat` for it after all. Sorry for the churn. We probably should have discussed the design in a bug before you implemented it.

From your original message, I'm guessing that you don't have a need for `wValidBitsPerSample` or `dwChannelMask`. It sounds like you're receiving these wavs from a source that happens to write them in extensible format, and you just want to be able to read them like any other wav and get the samples. Is that the case?

Assuming that's true, here is my proposal:
- We don't add any values to `WavFormat` or add any fields to `Wav`
- We support `WAVE_FORMAT_EXTENSIBLE` for reading, but not for writing
- The only field of the extensible chunk we really care about is `SubFormat`. We decode the `WavFormat` from this field, and stick it in `Wav.format`.
- If the `wValidBitsPerSample` field doesn't match `wBitsPerSample`, we throw an exception that tells the user to file a bug and send us their wav file. That way we can add support for it once we have a real wav to take a look at, and I can figure out how it's encoded
- We ignore the `dwChannelMask` field until a user needs it
- In other words, we'll just convert the extensible format into the ordinary format.

Another question: how did you get the test sine wavs? Did you use `Wav.write()` to generate them, or get them from a separate source?

dtruong27 commented:
Thanks a lot for another review! No worry, it was not too much work for me, I should have discussed before implementing.

Now I reverted some changes and implemented based on your proposals. Please check again, as I am not sure if I understood completely your ideas.

The sine wav files were generated in a `dart` script using this package directly. I did test them and they could be read by Audacity and Quicktime player.

liamappelbe commented on lib/wav_file.dart:
```diff
@@ -89,14 +91,45 @@ class Wav {
+      if (cbSize == _kExCbSize) {
+        final validBitsPerSample = byteReader.readUint16();
+        if (validBitsPerSample != bitsPerSample) {
+          throw FormatException(
```
Make this a `UnimplementedError` and add `"Please file a bug on package:wav"`

dtruong27 replied:
Nothing to sorry, we are not in a rush :)
Thanks for having a look again. I tried applying your suggestion.

liamappelbe reviewed (commented):
Sorry for the delay. I'm travelling atm. Looks pretty good tho. Just one more small comment.

liamappelbe reviewed (approved):
Looks good!
